### ISSUE TITLE
Target::Pipe doesn't work

### ISSUE THREAD

**Issue by oxalica** [env-logger-rs/env_logger#208]
In #204, we introduce `Target::Pipe` to redirect log messages into custom writer. And it's included in 0.8.4 now.
But when I try [the example in that PR](https://github.com/env-logger-rs/env_logger/pull/204/files#diff-24824a94b419dd85f77d13b3858f55f97857a3093a7a862d180d54a83786ad6b), it doesn't print the message `from pipe` at all, and not even call `WriteAdapter::write`. The `log::error` message just go straight to stderr.
How to make it catch all log message into my custom writer object?

**Comment by TilCreator** [comment]
I also just noticed that. I read the code (of #204), but never actually tested it... I kind of just assumed that it would work, because it didn't change much of my code (#195).
I'm currently looking into it (I also already found at least one error, but it's just in the documentation)

**Comment by jplatte** [comment]
Thanks for looking into it, and sorry for breaking things.

**Comment by TilCreator** [comment]
I have found the error in [`fmt::writer::termcolor::extern_impl::BufferWriter`](https://github.com/env-logger-rs/env_logger/pull/204/commits/0f82db494fdcd3737fc489b69907875c4f316517#diff-c2eb0d1b5b63c0f6315a12f1a11191d82f9a45306ebc0ca060b5757871cb5884L72)
The change removed the `target_pipe` which is good, less convoluted.
In [`fmt::writer::termcolor::extern_impl::BufferWriter::print`](https://github.com/env-logger-rs/env_logger/pull/204/commits/0f82db494fdcd3737fc489b69907875c4f316517#diff-c2eb0d1b5b63c0f6315a12f1a11191d82f9a45306ebc0ca060b5757871cb5884L127) the if for the pipe was removed and integrated into the match a few lines down, but that only works if there is a `test_target`, because `BufferWriter::inner` (a `termcolor::BufferWriter`) is unable to write to the pipe.

**Comment by TilCreator** [comment]
A easy solution would be to change BufferWriter::pipe to always set `test_target`, but that feels quite hacky...

**Comment by jplatte** [comment]
I think we should no longer unconditionally hold that `BufferWriter`, it doesn't make sense for the pipe target.

> it also removes color, so not really a solution

Removes color in general, or when using a `Pipe` writer? For the latter, I would *expect* colorization to be turned off. You could print the escape sequences, but that's unix-specific and would lead to very ugly output if it's not expected on the "other side" of the pipe.

**Comment by TilCreator** [comment]
I explicitly enabled colors, else the default for pipes would be to disable color.

**Comment by TilCreator** [comment]
Any idea other than to reimplement some version of the old `target_pipe`?

**Comment by TilCreator** [comment]
I will try the idea without the `BufferWriter`

**Comment by TilCreator** [comment]
> I think we should no longer unconditionally hold that BufferWriter, it doesn't make sense for the pipe target.

Something like this?
https://github.com/TilCreator/env_logger/commit/fbeebcf1081815b61568fb410b8dbb9747c57e0b

**Comment by tubzby** [comment]
Same problem here.

**Comment by Polochon-street** [comment]
Can report having the same problem - that would be a really awesome addition if that worked though :smile:

**Comment by TilCreator** [comment]
Ups, it has been a _few_ days...
Sry, I opened #211 which fixes this (but it's not perfect).

**Comment by mainrs** [comment]
II do also think that colors disabled makes more sense for pipes (by default). We could think about adding some kind of way to enable them back again.

**Comment by TilCreator** [comment]
It is already disabled by default for pipes, the example just enables them to show that it's possible.

**Comment by harlanc** [comment]
Is this problem fixed now? Both 0.8.4 and 0.9 versions have problems..

**Comment by jplatte** [comment]
Seems like #211 would fix this.

**Comment by harlanc** [comment]
when will this patch be merged and released? @TilCreator

**Comment by TilCreator** [comment]
No idea, but I hope soon

**Comment by tubzby** [comment]
Any news on this?
I think this is a critical issue because I can't imagine a process without a log file in the production environment.

**Comment by jplatte** [comment]
#217 contains an incomplete fix for this. I'm not sure it can really be completed without breaking changes though.

**Comment by TilCreator** [comment]
I'm quite busy right now, so I will sadly not be of much help, sry (also my **non critical** applications currently work fine with #211)

**Comment by yihuaf** [comment]
@jplatte Any update? Can we re-open #211 if it fix the issue? Being able to log to file is a critical feature.

**Comment by jplatte** [comment]
I think you'd be better off just using a different logger like [simplelog](https://docs.rs/simplelog) if you need file output. If you really want it work in `env_logger`, you could try working on #217.

**Comment by hasezoey** [comment]
just encountered this issue as well, it is confusing that this feature (`Target::Pipe`) does not work (silently) without having documentation mentioning that it does not work

also is there any progress on #217 to be ready for merge?

**Comment by jplatte** [comment]
Yeah, I suppose some docs plus a `deprecated` attribute (because that is the only way you can make usage of it warn, AFAIK) would be a good interim solution. There has been no progress on #217, help wanted.

**Comment by SolraBizna** [comment]
I just burned some debugging time trying to figure out why my custom pipe writer wasn't working, only to stumble upon this issue. There has, according to my testing, never been a published version of this crate where this feature worked. I am vexed!

"Make your own logger" isn't a good solution, because then we give up `env_logger`'s flagship (and titular) feature, which is its simple and powerful system by which the *user* specifies which log messages they're interested in.

I intend to put my money where my mouth is and throw myself at fixing this today or tomorrow.

**Comment by niklasha** [comment]
Heh, I too burnt hours debugging this, and made a fix of my own before checking the issues.  This really ought to be fixed, either by correcting the bug (easy enough), or.. if it is easier, just by releasing a new version with fixed documentation (where it is noted that Pipe does not work unless in test context).  If it is documented as working, people will continue to trip on this.

**Comment by niklasha** [comment]
and for the record: this was how I fixed it locally: https://github.com/niklasha/env_logger/tree/issue-208-pipe-fix
This is basically just making the uncolored case be used always with Pipe, be it in testing context or not.  I know there is already a rejected proposal, but i thought I'd at least share what I did, should anyone like it.

**Comment by whirm** [comment]
I've also been a victim of this. I spent hours debugging this until I realized not even the example worked.

**Comment by walkie** [comment]
@niklasha:
> this was how I fixed it locally: https://github.com/niklasha/env_logger/tree/issue-208-pipe-fix
> This is basically just making the uncolored case be used always with Pipe, be it in testing context or not. I know there is already a rejected proposal, but i thought I'd at least share what I did, should anyone like it.

I've read through all of the various issues and pull requests, and there seems to be a consensus that disabling color for pipe targets (at least initially) is acceptable.

For example, this comment from @mainrs:
> I do also think that colors disabled makes more sense for pipes (by default). We could think about adding some kind of way to enable them back again.

Is there any reason that @niklasha's patch is not a valid solution to this issue?

If not, I'm happy to create a small PR updating the documentation to clarify that this feature is only supported in test contexts. It's definitely confusing in its current state, as the last few commenters here can attest. :-)

**Comment by jplatte** [comment]
That approach sounds reasonable, I'd be happy about a PR.

**Comment by SolraBizna** [comment]
[My PR over at the termcolor repo](https://github.com/BurntSushi/termcolor/pull/62) will make a simple path for enabling color for pipe targets down the road. But it doesn't do us any good until it's on crates.io.

**Comment by TilCreator** [comment]
@walkie I thinks that color output should be disabled be default, but I think it still is an important feature, for example I use the color output to combine the indicatif progress bar and pretty env logger in https://gitlab.com/TilCreator/rustycomicscraper/-/blob/master/src/main.rs

**Comment by jplatte** [comment]
I've published version 0.9.1 that deprecates `Target::Pipe`.

**Comment by niklasha** [comment]
> I've published version 0.9.1 that deprecates `Target::Pipe` because nobody is working on fixing it.

I offer to fix it, if I only knew what the requirements on a solution are.  I have already devised a solution that "works for me" above: https://github.com/env-logger-rs/env_logger/issues/208#issuecomment-1173592691 but I understand you have issues with it?

**Comment by jplatte** [comment]
I already told you that I would be happy to review a PR with your patch, but I don't think you created one?

**Comment by niklasha** [comment]
Eh, you did?  I must have misunderstood then.  I will prepare a PR.

**Comment by epage** [comment]
@jplatte did 84b701dac94e4135f59f069a8c155bb9b7922788 resolve this Issue?

**Comment by jplatte** [comment]
It should have, yes.
